### PR TITLE
Add ResNet & refactor

### PR DESCRIPTION
**Pull request by jinPrelude**
add resnet and put resnet.py & cnn.py into rl_algorithms/common.networks.backbones directory.

### CODE REVIEW

**Comment by jinPrelude** [comment]
resnetiqn_mint
benchmark comparison of non-registry resnet iqn & registry-structured resnet iqn. Mint color is registry_structured resnet iqn.

File: rl_algorithms/__init__.py [hunk 1]
```diff
@@ -1,8 +1,10 @@
+from rl_algorithms.common.networks.backbones.cnn import CNN
+from rl_algorithms.common.networks.backbones.resnet import ResNet
```
Comment (MrSyee) [inline rl_algorithms/__init__.py:2]: should change `from rl_algorithms.common.networks.backbones import CNN, ResNet` using `__init__.py`

File: configs/pong_no_frameskip_v4/iqnResNet.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            expansion=1,
+            channel_compression=4,  # output channel // channel_compression in last conv layer
+        ),
+    ),
+    head=dict(
+        type="IQNMLP",
+        configs=dict(
+            # NoisyNet
+            use_noisy_net=True,
+            std_init=0.5,
```
Comment (MrSyee) [inline configs/pong_no_frameskip_v4/iqnResNet.py:51]: should move below

File: rl_algorithms/__init__.py (hunk 1 above)
Comment (MrSyee) [inline rl_algorithms/__init__.py:2]: should change from rl_algorithms.common.networks.backbones import CNN, ResNet using __init__.py

File: configs/pong_no_frameskip_v4/iqnResNet.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+"""Config for DQN on Pong-No_FrameSkip-v4.
```
Comment (MrSyee) [inline configs/pong_no_frameskip_v4/iqnResNet.py:1]: should change filename from `iqnResNet.py` to `iqn_resnet.py`

File: rl_algorithms/common/networks/base_network.py [hunk 4]
```diff
@@ -2,8 +2,9 @@
 import torch.nn as nn
 
 from rl_algorithms.common.helper_functions import identity
-from rl_algorithms.common.networks.cnn import CNN
-from rl_algorithms.common.networks.mlp import FlattenMLP
+from rl_algorithms.common.networks.backbones.cnn import CNN
+from rl_algorithms.common.networks.backbones.resnet import ResNet
```
Comment (MrSyee) [inline rl_algorithms/common/networks/base_network.py:6]: same comment

File: tests/test_cnn_cfg.py [hunk 5]
```diff
@@ -1,19 +1,36 @@
 import torch
 import torch.nn as nn
 
+from rl_algorithms.common.networks.backbones.cnn import CNN
+from rl_algorithms.common.networks.backbones.resnet import ResNet
```
Comment (MrSyee) [inline tests/test_cnn_cfg.py:5]: same comment: https://github.com/medipixel/rl_algorithms/pull/209/files#r411047082

File: rl_algorithms/common/networks/base_network.py [hunk 6]
```diff
@@ -46,11 +47,12 @@ def forward_(self, x: torch.Tensor, n_tau_samples: int = None):
         return x
 
 
-def calculate_fc_input_size(state_dim: tuple, cnn: ConfigDict):
+def calculate_fc_input_size(state_dim: tuple, backbone_cfg: ConfigDict):
     """calculate fc input size according to the shape of cnn"""
```
Comment (MrSyee) [inline rl_algorithms/common/networks/base_network.py:51]: - Capitalize the first letter
- Put a period

File: rl_algorithms/common/networks/backbones/resnet.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+
+from rl_algorithms.registry import BACKBONES
+from rl_algorithms.utils.config import ConfigDict
+
+device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+
+class BasicBlock(nn.Module):
```
Comment (MrSyee) [inline rl_algorithms/common/networks/backbones/resnet.py:18]: should add simple documentation

File: rl_algorithms/common/networks/base_network.py [hunk 8]
```diff
@@ -46,11 +47,12 @@ def forward_(self, x: torch.Tensor, n_tau_samples: int = None):
         return x
 
 
-def calculate_fc_input_size(state_dim: tuple, cnn: ConfigDict):
+def calculate_fc_input_size(state_dim: tuple, backbone_cfg: ConfigDict):
```
Comment (MrSyee) [inline rl_algorithms/common/networks/base_network.py:50]: I think this method should move into `BaseNetwork` due to removing if condition.

File: configs/pong_no_frameskip_v4/iqnResNet.py (hunk 2 above)
Comment (jinPrelude) [reply configs/pong_no_frameskip_v4/iqnResNet.py]: resolved in commit 2cad1a7!!

File: rl_algorithms/__init__.py (hunk 1 above)
Comment (jinPrelude) [reply rl_algorithms/__init__.py]: resolved in commit 2cad1a7!!

File: configs/pong_no_frameskip_v4/iqnResNet.py (hunk 3 above)
Comment (jinPrelude) [reply configs/pong_no_frameskip_v4/iqnResNet.py]: resolved in commit 2cad1a7!!

File: tests/test_cnn_cfg.py (hunk 5 above)
Comment (jinPrelude) [reply tests/test_cnn_cfg.py]: resolved in commit 2cad1a7!!

File: rl_algorithms/common/networks/base_network.py (hunk 4 above)
Comment (jinPrelude) [reply rl_algorithms/common/networks/base_network.py]: resolved in commit 2cad1a7!!

File: rl_algorithms/common/networks/base_network.py (hunk 6 above)
Comment (jinPrelude) [reply rl_algorithms/common/networks/base_network.py]: resolved in commit 2cad1a7!!

File: rl_algorithms/common/networks/backbones/resnet.py (hunk 7 above)
Comment (jinPrelude) [reply rl_algorithms/common/networks/backbones/resnet.py:18]: resolved in commit 2cad1a7!!

File: rl_algorithms/common/networks/base_network.py (hunk 8 above)
Comment (jinPrelude) [reply rl_algorithms/common/networks/base_network.py]: resolved in commit 2cad1a7!!

**Pull request merged**